### ISSUE TITLE
734ms regression in cold-view-nav-start on 7/14 nightly

### ISSUE THREAD

**Issue by mcomella** [mozilla-mobile/fenix#20934]
![image](https://user-images.githubusercontent.com/759372/130148281-29324e4e-5969-4fb4-b30d-e131ab82381c.png)

We regressed 734ms on my Moto G5: let's find the root cause.

**Comment by mcomella** [comment]
I took two profiles:
- (unaffected) 7/12 nightly: https://share.firefox.dev/3DeDLOT
- (affected) 7/20 nightly: https://share.firefox.dev/3D0QJj7

In the parent process for 7/20:
- the time to the 3rd dom navigation start (which the profiles are clamped to) is 0.9s longer from the start of the profile
- the jank markers are much longer
- the `parseManifest` block is much longer:

![image](https://user-images.githubusercontent.com/759372/130301636-9682dbe3-41b2-4a69-bd07-1d113dbdfb39.png)

It goes from 482ms to 1330ms. It's unclear what the root cause is.

**Comment by Rob--W** [comment]
FYI: I'm an engineer that works on WebExtensions framework in Firefox.

From the profile and the graph, it appears that quite some time is spent on parsing the values of the `permissions` key in the manifests of some extensions. On the 14th, the search and ads extensions in A-C were updated with a few hundred additional entries in the permissions section of manifest.json: https://github.com/mozilla-mobile/android-components/pull/10541

The implementation that parses the permissions section in the manifest file is quite generic. If this is a significant perf issue we could look into optimizing it. I'm not familiar with how the profiles here are generated, so I need an explanation on how I can view and run the profiler to compare the difference between the current performance and the (new/improved) result.

Note: the result of these parsed manifests are cached (keyed by application version, extension version and language), so in practice users won't experience this worst case except during application upgrades and extension updates.

**Comment by mcomella** [comment]
> From the profile and the graph, it appears that quite some time is spent on parsing the values of the permissions key in the manifests of some extensions. On the 14th, the search and ads extensions in A-C were updated with a few hundred additional entries in the permissions section of manifest.json: mozilla-mobile/android-components#10541

I believe this is the cause after making a proof-of-concept. I verified I could reproduce the problem locally [1] then checked out 2040af6a5 (a commit from 7/20 with the regression), removed the items from the permissions list (e.g. [this](https://github.com/mozilla-mobile/android-components/blob/5d8f588f26f37688f02516a3b9ee0b7d52a689c8/components/feature/search/src/main/assets/extensions/search/manifest.template.json#L14-L208) & similar. [20934-permissions-changes-diff.txt](https://github.com/mozilla-mobile/fenix/files/7035483/20934-permissions-changes.txt)), and compared the benchmark between the two:
- 2040af6a5 with local ac 5630e1f65c: 2748ms. https://share.firefox.dev/388m5WA
- 2040af6a5 with local ac 5630e1f65c & permissions removed: 2346ms. https://share.firefox.dev/3B5BOlF

i.e. the regression was removed.

[1]: I locally checked out 2c1d0ea18, a commit from 7/12, and 2040af6a5, a commit from 7/20, and saw a regression (2405ms to 2763ms). Curiously, **this regression isn't as large as the one we saw in Nightly.**

### Solutions
Assuming this is the case, I think we have two options:
1. optimize the parsing code
2. reduce the number of permissions for this functionality

Let's figure out which one makes the most sense.

@Rob--W This is a significant regression: would you or your team be able to optimize the parsing code and what would the turn around time be? Unfortunately, I suspect this affects Beta so we have a limited amount of time.

@Mugurell https://github.com/mozilla-mobile/android-components/pull/10541 caused start up time to regress significantly: are those changes needed in fenix v92 (i.e. the next release) or is that something we could back out to give Rob--W more time? Alternatively, could we ship this feature with a smaller set of permissions in the manifest?

Sorry this is so last minute – our perf testing system has been in flux. :\

**Comment by mcomella** [comment]
> Note: the result of these parsed manifests are cached (keyed by application version, extension version and language), so in practice users won't experience this worst case except during application upgrades and extension updates.

@Rob--W When do these get cached? Do we have any data to know how often users on Android hit the cache? That might change how urgent this is.

For context, our current perf tests start the app, wait 5s, force stop the app (i.e. this is to make sure it's not a first run) and then start measuring start up performance.

**Comment by Mugurell** [comment]
Regarding the many new permissions we tried to specialize the needed ones for each website instead of using wildcards that would match any website. Mostly as a security measure.
Alternatively we could use wildcards if the permissions are to be requested only by the websites declared in the `content_scripts` block?

**Comment by Rob--W** [comment]
> > Note: the result of these parsed manifests are cached (keyed by application version, extension version and language), so in practice users won't experience this worst case except during application upgrades and extension updates.
> 
> @Rob--W When do these get cached?

... actually, that caching does not happen for built-in extensions on mobile. Usually, the `Extension` class takes care of it, but the `ensureBuiltIn` implementation instantiates an instance of the `ExtensionData` class instead: https://searchfox.org/mozilla-central/rev/7bafa9b9c2156018ec1d410194f9bf4b5b23e77f/mobile/android/modules/geckoview/GeckoViewWebExtension.jsm#658,660

> Do we have any data to know how often users on Android hit the cache? That might change how urgent this is.

This data is not tracked. But as I mentioned above, the implementation for built-ins does not use caching after all.




> Regarding the many new permissions we tried to specialize the needed ones for each website instead of using wildcards that would match any website. Mostly as a security measure.
> @Rob--W I don't really know how these permissions are parsed and matched but I think they might lead to an O(n²) complexity? which maybe we can avoid if we are to match `<all_urls>` if we know those permissions are requested only by the websites declared in the `content_scripts` block and this would still be secure, not leak anything, etc?

Both extensions request loads of permissions, but these appear to all be unnecessary for the extensions to function:
- https://github.com/Mugurell/android-components/blob/6f85faeb33b3e351effc02d1bca1cacc55fe4820/components/feature/search/src/main/assets/extensions/ads/manifest.template.json#L217-L411
- https://github.com/Mugurell/android-components/blob/6f85faeb33b3e351effc02d1bca1cacc55fe4820/components/feature/search/src/main/assets/extensions/search/manifest.template.json#L218-L412

The same patterns are already repeated in the `include_globs` section. Incidentally, parsing an `content_scripts[*].include_globs` entry is slightly less expensive than `content_scripts[*].matches`, because `include_globs` is just defined to be a string, whereas additional validation is performed on `matches` ([definition of the types here](https://searchfox.org/mozilla-central/rev/7bafa9b9c2156018ec1d410194f9bf4b5b23e77f/toolkit/components/extensions/schemas/manifest.json#603-604,606-607,615-616,618)).

I think that the perf hit would be avoided without functionality degradation by removing the lines that I mentioned.

**Comment by Rob--W** [comment]
I filed a bug in GeckoView::Extensions to address the identified inefficiency in GeckoView's implementation of the built-in extension check: https://bugzilla.mozilla.org/show_bug.cgi?id=1727272

**Comment by Mugurell** [comment]
> Both extensions request loads of permissions, but these appear to all be unnecessary for the extensions to function:
    * https://github.com/Mugurell/android-components/blob/6f85faeb33b3e351effc02d1bca1cacc55fe4820/components/feature/search/src/main/assets/extensions/ads/manifest.template.json#L217-L411
     * https://github.com/Mugurell/android-components/blob/6f85faeb33b3e351effc02d1bca1cacc55fe4820/components/feature/search/src/main/assets/extensions/search/manifest.template.json#L218-L412
> 
> I think that the perf hit would be avoided without functionality degradation by removing the lines that I mentioned.

@Rob--W So we can just remove those lines, no need to also have `<all_urls>` matcher in their place?

**Comment by Rob--W** [comment]
> > Both extensions request loads of permissions, but these appear to all be unnecessary for the extensions to function:
> > * https://github.com/Mugurell/android-components/blob/6f85faeb33b3e351effc02d1bca1cacc55fe4820/components/feature/search/src/main/assets/extensions/ads/manifest.template.json#L217-L411
> > * https://github.com/Mugurell/android-components/blob/6f85faeb33b3e351effc02d1bca1cacc55fe4820/components/feature/search/src/main/assets/extensions/search/manifest.template.json#L218-L412
> > I think that the perf hit would be avoided without functionality degradation by removing the lines that I mentioned.
> 
> @Rob--W So we can just remove those lines, no need to also have `<all_urls>` matcher in their place?
> The intention there was to limit the permissions to only the websites we are interested in but I understand that having the websites declared in `include_globs` is enough.

Yep remove without replacement as the content script does not do anything that requires host permissions - https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/manifest.json/permissions#host_permissions

The pattern in `matches` is primarily responsible for determining which set of sites to match at first, and then `include_globs` imposes further restrictions on the set of matches URLs. The host permissions seen by extensions is the combination of all host permissions in `permissions` plus the patterns in `content_scripts[*].matches`.

**Comment by Mugurell** [comment]
Did a light testing and I see out extensions working as before. Will open a PR on AC to remove those lines.

**Comment by mcomella** [comment]
Measuring locally, the PR appears to have addressed the issue. Here's the analysis: https://github.com/mozilla-mobile/android-components/pull/10872#issuecomment-904858813 **We should confirm the regression is addressed on nightly before closing this bug.**

See https://github.com/mozilla-mobile/android-components/pull/10874 for the backport.
